src/core/execute-serialize.c

```diff
@@ -0,0 +1,4146 @@
+        if (r < 0)
+                return r;
+
+        r = serialize_fd_array_or_set(f, fds, fds_array, n_fds_array, "exec-parameters-exec-fd", p->exec_fd);
+        if (r < 0)
+                return r;
+
+        r = serialize_fd_array_or_set(f, fds, fds_array, n_fds_array, "exec-parameters-bpf-outer-map-fd", p->bpf_outer_map_fd);
+        if (r < 0)
+                return r;
+
+        r = serialize_item(f, "exec-parameters-notify-socket", p->notify_socket);
+        if (r < 0)
+                return r;
+
+        LIST_FOREACH(open_files, file, p->open_files) {
+                _cleanup_free_ char *ofs = NULL;
+
+                r = open_file_to_string(file, &ofs);
+                if (r < 0)
+                        return r;
+
+                r = serialize_item(f, "exec-parameters-open-file", ofs);
+                if (r < 0)
+                        return r;
+        }
+
+        r = serialize_item(f, "exec-parameters-default-smack-process-label", p->default_smack_process_label);
+        if (r < 0)
+                return r;
+
+        r = serialize_fd_array_or_set(f, fds, fds_array, n_fds_array, "exec-parameters-user-lookup-fd", p->user_lookup_fd);
+        if (r < 0)
+                return r;
+
+        r = serialize_strv(f, "exec-parameters-files-env", p->files_env);
+        if (r < 0)
+                return r;
+
+        fputc('\n', f); /* End marker */
+
+        return 0;
+}
+
+static int exec_parameters_deserialize(ExecParameters *p,
+                FILE *f,
+                FDSet *fds,
+                int *fds_array,
+                size_t n_fds_array) {
+
+        int r;
+
+        assert(p);
+        assert(f);
+        assert(!(fds && fds_array));
+        assert(!!fds_array == !!n_fds_array);
+
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+                const char *val, *l;
+
+                r = read_line(f, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read serialization line: %m");
+                if (r == 0)
+                        break;
+
+                l = strstrip(line);
+                if (isempty(l)) /* end marker */
+                        break;
+
+                if ((val = startswith(l, "exec-parameters-environment="))) {
+                        r = deserialize_strv(&p->environment, val);
+                        if (r < 0)
+                                return r;
+                } else if ((val = startswith(l, "exec-parameters-n-socket-fds="))) {
+                        r = safe_atozu(val, &p->n_socket_fds);
+                        if (r < 0)
+                                return r;
+                } else if ((val = startswith(l, "exec-parameters-n-storage-fds="))) {
+                        r = safe_atozu(val, &p->n_storage_fds);
+                        if (r < 0)
+                                return r;
+                } else if ((val = startswith(l, "exec-parameters-fds="))) {
```

Comment:
This one is a bit of a pickle, as there are multiple possible issues.

1) With a large enough value for `exec-parameters-n-{storage,socket)-fds=` this will trigger multiple runtime errors:

```
Line: exec-parameters-n-storage-fds=999999999999  
Line: exec-parameters-fds=                                                                                             
=================================================================                                                                                                                                                                              
==1084343==ERROR: AddressSanitizer: requested allocation size 0x3a352943ffc (0x3a352945000 after adjustments for alignment, red zones etc.) exceeds maximum supported size of 0x10000000000 (thread T0)                                        
    #0 0x7f998f0d92ff in malloc (/lib64/libasan.so.8+0xd92ff) (BuildId: dc689b05ca2577037af24700212bb5cce1f91c8a)      
    #1 0x7f998e13456a in malloc_multiply ../src/basic/alloc-util.h:120                                                 
    #2 0x7f998e15096c in exec_parameters_deserialize ../src/core/execute-serialize.c:1673                              
    #3 0x7f998e17987c in exec_deserialize ../src/core/execute-serialize.c:4167                                         
    #4 0x40328b in LLVMFuzzerTestOneInput ../src/core/fuzz-execute-serialize.c:62                                                                                                                                                              
    #5 0x403d35 in main ../src/fuzz/fuzz-main.c:50                                                                                                                                                                                             
    #6 0x7f998b249b49 in __libc_start_call_main (/lib64/libc.so.6+0x27b49) (BuildId: 245240a31888ad5c11bbc55b18e02d87388f59a9)                                                                                                                 
    #7 0x7f998b249c0a in __libc_start_main_alias_2 (/lib64/libc.so.6+0x27c0a) (BuildId: 245240a31888ad5c11bbc55b18e02d87388f59a9)                                                                                                              
    #8 0x402374 in _start (/home/mrc0mmand/repos/@systemd/systemd/build-san/fuzz-execute-serialize+0x402374) (BuildId: eeba9a7a26ddb9a435adebf4eba0e129b287521e)                                                                               
```

or 

```
Line: exec-parameters-n-storage-fds=9999999                                                                                                                                                                                                    
Line: exec-parameters-fds=                                                                                                                                                                                                                     
Failed to serialize parameters: Too many open files                                                                    
Assertion 'close_nointr(fd) != -EBADF' failed at src/basic/fd-util.c:75, function safe_close(). Aborting.   
```

2) If you order the serialization directives in a particular order, you can trigger a heap-buffer-overflow. I.e. with:

```
exec-unit-type=mount



exec-parameters-n-storage-fds=484
exec-parameters-fds=484
exec-parameters-n-storage-fds=4484
```

you'll end up with:

```
Line: exec-parameters-fds=484
FD 484 not in fdset.
Failed to deserialize exec-parameters-fds=484 value: 484, ignoring.
Line: exec-parameters-n-storage-fds=4484
=================================================================
==1085311==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x61d000000810 at pc 0x7f58feb4e19c bp 0x7fff657dbf70 sp 0x7fff657dbf68
READ of size 4 at 0x61d000000810 thread T0
    #0 0x7f58feb4e19b in exec_parameters_serialize ../src/core/execute-serialize.c:1484
    #1 0x7f58feb79055 in exec_serialize ../src/core/execute-serialize.c:4124
    #2 0x403459 in LLVMFuzzerTestOneInput ../src/core/fuzz-execute-serialize.c:65
    #3 0x403d35 in main ../src/fuzz/fuzz-main.c:50
    #4 0x7f58fbc49b49 in __libc_start_call_main (/lib64/libc.so.6+0x27b49) (BuildId: 245240a31888ad5c11bbc55b18e02d87388f59a9)
    #5 0x7f58fbc49c0a in __libc_start_main_alias_2 (/lib64/libc.so.6+0x27c0a) (BuildId: 245240a31888ad5c11bbc55b18e02d87388f59a9)
    #6 0x402374 in _start (/home/mrc0mmand/repos/@systemd/systemd/build-san/fuzz-execute-serialize+0x402374) (BuildId: eeba9a7a26ddb9a435adebf4eba0e129b287521e)

0x61d000000810 is located 0 bytes after 1936-byte region [0x61d000000080,0x61d000000810)
allocated by thread T0 here:
    #0 0x7f58ffad92ff in malloc (/lib64/libasan.so.8+0xd92ff) (BuildId: dc689b05ca2577037af24700212bb5cce1f91c8a)
    #1 0x7f58feb3456a in malloc_multiply ../src/basic/alloc-util.h:120
    #2 0x7f58feb5096c in exec_parameters_deserialize ../src/core/execute-serialize.c:1673
    #3 0x7f58feb7987c in exec_deserialize ../src/core/execute-serialize.c:4167
    #4 0x40328b in LLVMFuzzerTestOneInput ../src/core/fuzz-execute-serialize.c:62
    #5 0x403d35 in main ../src/fuzz/fuzz-main.c:50
    #6 0x7f58fbc49b49 in __libc_start_call_main (/lib64/libc.so.6+0x27b49) (BuildId: 245240a31888ad5c11bbc55b18e02d87388f59a9)
    #7 0x7f58fbc49c0a in __libc_start_main_alias_2 (/lib64/libc.so.6+0x27c0a) (BuildId: 245240a31888ad5c11bbc55b18e02d87388f59a9)
    #8 0x402374 in _start (/home/mrc0mmand/repos/@systemd/systemd/build-san/fuzz-execute-serialize+0x402374) (BuildId: eeba9a7a26ddb9a435adebf4eba0e129b287521e)

SUMMARY: AddressSanitizer: heap-buffer-overflow ../src/core/execute-serialize.c:1484 in exec_parameters_serialize
Shadow bytes around the buggy address:
  0x61d000000580: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x61d000000600: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x61d000000680: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x61d000000700: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x61d000000780: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x61d000000800: 00 00[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x61d000000880: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x61d000000900: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x61d000000980: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x61d000000a00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x61d000000a80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1085311==ABORTING
```

since we allocated the fds array according to the first value, but didn't update it after receiving the second one.

Reply:
I've added an upper limit, given we know how many open FDs we can have at any given time that should be enough